File: integration-tests/TestAPI-bindings/kotlin/src/test/kotlin/com/cricut/testapi/RangeTests.kt
package com.cricut.testapi

import com.cricut.fishyjoes.runtime.SwiftClosedRange
import com.cricut.fishyjoes.runtime.SwiftRange
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

internal class RangeTests {
    @Test
    fun testRangeValues() {
        assertEquals(Ranges.uInt8Range, SwiftRange(UByte.MIN_VALUE, UByte.MAX_VALUE))
        assertEquals(Ranges.uInt16Range, SwiftRange(UShort.MIN_VALUE, UShort.MAX_VALUE))
        assertEquals(Ranges.uInt32Range, SwiftRange(UInt.MIN_VALUE, UInt.MAX_VALUE))
        assertEquals(Ranges.uInt64Range, SwiftRange(ULong.MIN_VALUE, ULong.MAX_VALUE))
        assertEquals(Ranges.uIntRange, SwiftRange(ULong.MIN_VALUE, ULong.MAX_VALUE))
        assertEquals(Ranges.int8Range, SwiftRange(Byte.MIN_VALUE, Byte.MAX_VALUE))
        assertEquals(Ranges.int16Range, SwiftRange(Short.MIN_VALUE, Short.MAX_VALUE))
        assertEquals(Ranges.int32Range, SwiftRange(Int.MIN_VALUE, Int.MAX_VALUE))
        assertEquals(Ranges.int64Range, SwiftRange(Long.MIN_VALUE, Long.MAX_VALUE))
        assertEquals(Ranges.intRange, SwiftRange(Long.MIN_VALUE, Long.MAX_VALUE))
    }

    @Test
    fun testRangeEcho() {
        assertEquals(Ranges.echoUInt8Range(Ranges.uInt8Range), Ranges.uInt8Range)
        assertEquals(Ranges.echoUInt16Range(Ranges.uInt16Range), Ranges.uInt16Range)
        assertEquals(Ranges.echoUInt32Range(Ranges.uInt32Range), Ranges.uInt32Range)
        assertEquals(Ranges.echoUInt64Range(Ranges.uInt64Range), Ranges.uInt64Range)
        assertEquals(Ranges.echoUIntRange(Ranges.uIntRange), Ranges.uIntRange)
        assertEquals(Ranges.echoInt8Range(Ranges.int8Range), Ranges.int8Range)
        assertEquals(Ranges.echoInt16Range(Ranges.int16Range), Ranges.int16Range)
        assertEquals(Ranges.echoInt32Range(Ranges.int32Range), Ranges.int32Range)
        assertEquals(Ranges.echoInt64Range(Ranges.int64Range), Ranges.int64Range)
        assertEquals(Ranges.echoIntRange(Ranges.intRange), Ranges.intRange)
    }

    @Test
    fun testClosedRangeValues() {
        assertEquals(ClosedRanges.uInt8Range, SwiftClosedRange(UByte.MIN_VALUE, UByte.MAX_VALUE))
        assertEquals(ClosedRanges.uInt16Range, SwiftClosedRange(UShort.MIN_VALUE, UShort.MAX_VALUE))
        assertEquals(ClosedRanges.uInt32Range, SwiftClosedRange(UInt.MIN_VALUE, UInt.MAX_VALUE))
        assertEquals(ClosedRanges.uInt64Range, SwiftClosedRange(ULong.MIN_VALUE, ULong.MAX_VALUE))
        assertEquals(ClosedRanges.uIntRange, SwiftClosedRange(ULong.MIN_VALUE, ULong.MAX_VALUE))
        assertEquals(ClosedRanges.int8Range, SwiftClosedRange(Byte.MIN_VALUE, Byte.MAX_VALUE))
        assertEquals(ClosedRanges.int16Range, SwiftClosedRange(Short.MIN_VALUE, Short.MAX_VALUE))
        assertEquals(ClosedRanges.int32Range, SwiftClosedRange(Int.MIN_VALUE, Int.MAX_VALUE))
        assertEquals(ClosedRanges.int64Range, SwiftClosedRange(Long.MIN_VALUE, Long.MAX_VALUE))
        assertEquals(ClosedRanges.intRange, SwiftClosedRange(Long.MIN_VALUE, Long.MAX_VALUE))
        assertEquals(ClosedRanges.floatRange, SwiftClosedRange(-Float.MAX_VALUE, Float.MAX_VALUE))
        assertEquals(ClosedRanges.doubleRange, SwiftClosedRange(-Double.MAX_VALUE, Double.MAX_VALUE))
        assertEquals(ClosedRanges.stringRange, SwiftClosedRange("A", "Z"))
    }

    @Test
    fun testClosedRangeEcho() {
        assertEquals(ClosedRanges.echoUInt8Range(ClosedRanges.uInt8Range), ClosedRanges.uInt8Range)
        assertEquals(ClosedRanges.echoUInt16Range(ClosedRanges.uInt16Range), ClosedRanges.uInt16Range)
        assertEquals(ClosedRanges.echoUInt32Range(ClosedRanges.uInt32Range), ClosedRanges.uInt32Range)
        assertEquals(ClosedRanges.echoUInt64Range(ClosedRanges.uInt64Range), ClosedRanges.uInt64Range)
        assertEquals(ClosedRanges.echoUIntRange(ClosedRanges.uIntRange), ClosedRanges.uIntRange)
        assertEquals(ClosedRanges.echoInt8Range(ClosedRanges.int8Range), ClosedRanges.int8Range)
        assertEquals(ClosedRanges.echoInt16Range(ClosedRanges.int16Range), ClosedRanges.int16Range)
        assertEquals(ClosedRanges.echoInt32Range(ClosedRanges.int32Range), ClosedRanges.int32Range)
        assertEquals(ClosedRanges.echoInt64Range(ClosedRanges.int64Range), ClosedRanges.int64Range)
        assertEquals(ClosedRanges.echoIntRange(ClosedRanges.intRange), ClosedRanges.intRange)
        assertEquals(ClosedRanges.echoFloatRange(ClosedRanges.floatRange), ClosedRanges.floatRange)
        assertEquals(ClosedRanges.echoDoubleRange(ClosedRanges.doubleRange), ClosedRanges.doubleRange)
        assertEquals(ClosedRanges.echoStringRange(ClosedRanges.stringRange), ClosedRanges.stringRange)
    }

    @Test
    fun testClosedRangeConversion() {
        val ub = 1.toUByte()..2.toUByte()
        val us = 1.toUShort()..2.toUShort()
        val ui = 1.toUInt()..2.toUInt()
        val uj = 1.toULong()..2.toULong()
        val b = 1.toByte()..2.toByte()
        val s = 1.toShort()..2.toShort()
        val i = 1.toInt()..2.toInt()
        val j = 1.toLong()..2.toLong()
        val f = 1.0F..2.0F
        val d = 1.0..2.0
        val l = "A".."Z"

        assertEquals(ClosedRanges.echoUInt32Range(SwiftClosedRange(ub)), SwiftClosedRange(ub))
        assertEquals(ClosedRanges.echoUInt32Range(SwiftClosedRange(us)), SwiftClosedRange(us))
        assertEquals(ClosedRanges.echoUInt32Range(SwiftClosedRange(ui)), SwiftClosedRange(ui))
        assertEquals(ClosedRanges.echoUIntRange(SwiftClosedRange(uj)), SwiftClosedRange(uj))
        assertEquals(ClosedRanges.echoInt32Range(SwiftClosedRange(b)), SwiftClosedRange(b))
        assertEquals(ClosedRanges.echoInt32Range(SwiftClosedRange(s)), SwiftClosedRange(s))
        assertEquals(ClosedRanges.echoInt32Range(SwiftClosedRange(i)), SwiftClosedRange(i))
        assertEquals(ClosedRanges.echoIntRange(SwiftClosedRange(j)), SwiftClosedRange(j))
        assertEquals(ClosedRanges.echoFloatRange(SwiftClosedRange(f)), SwiftClosedRange(f))
        assertEquals(ClosedRanges.echoDoubleRange(SwiftClosedRange(d)), SwiftClosedRange(d))
        assertEquals(ClosedRanges.echoStringRange(SwiftClosedRange(l)), SwiftClosedRange(l))
    }
}

File: integration-tests/TestAPI-bindings/c-sharp/Cricut.TestAPI.Tests/RangeTests.cs
using System;
using Xunit;
using Cricut.FishyJoesRuntime;

namespace Cricut.TestAPI.Tests {
    public class RangeTests {
        [Fact]
        void testRangeValues() {
            Assert.Equal(Ranges.UInt8Range, new SwiftRange<byte>(byte.MinValue, byte.MaxValue));
            Assert.Equal(Ranges.UInt16Range, new SwiftRange<ushort>(ushort.MinValue, ushort.MaxValue));
            Assert.Equal(Ranges.UInt32Range, new SwiftRange<uint>(uint.MinValue, uint.MaxValue));
            Assert.Equal(Ranges.UInt64Range, new SwiftRange<ulong>(ulong.MinValue, ulong.MaxValue));
            Assert.Equal(Ranges.UIntRange, new SwiftRange<nuint>(uint.MinValue, uint.MaxValue));
            Assert.Equal(Ranges.Int8Range, new SwiftRange<sbyte>(sbyte.MinValue, sbyte.MaxValue));
            Assert.Equal(Ranges.Int16Range, new SwiftRange<short>(short.MinValue, short.MaxValue));
            Assert.Equal(Ranges.Int32Range, new SwiftRange<int>(int.MinValue, int.MaxValue));
            Assert.Equal(Ranges.Int64Range, new SwiftRange<long>(long.MinValue, long.MaxValue));
            Assert.Equal(Ranges.IntRange, new SwiftRange<nint>(int.MinValue, int.MaxValue));
        }

        [Fact]
        void testRangeEcho() {
            Assert.Equal(Ranges.EchoUInt8Range(Ranges.UInt8Range), Ranges.UInt8Range);
            Assert.Equal(Ranges.EchoUInt16Range(Ranges.UInt16Range), Ranges.UInt16Range);
            Assert.Equal(Ranges.EchoUInt32Range(Ranges.UInt32Range), Ranges.UInt32Range);
            Assert.Equal(Ranges.EchoUInt64Range(Ranges.UInt64Range), Ranges.UInt64Range);
            Assert.Equal(Ranges.EchoUIntRange(Ranges.UIntRange), Ranges.UIntRange);
            Assert.Equal(Ranges.EchoInt8Range(Ranges.Int8Range), Ranges.Int8Range);
            Assert.Equal(Ranges.EchoInt16Range(Ranges.Int16Range), Ranges.Int16Range);
            Assert.Equal(Ranges.EchoInt32Range(Ranges.Int32Range), Ranges.Int32Range);
            Assert.Equal(Ranges.EchoInt64Range(Ranges.Int64Range), Ranges.Int64Range);
            Assert.Equal(Ranges.EchoIntRange(Ranges.IntRange), Ranges.IntRange);
        }

        [Fact]
        void testClosedRangeValues() {
            Assert.Equal(ClosedRanges.UInt8Range, new SwiftClosedRange<byte>(byte.MinValue, byte.MaxValue));
            Assert.Equal(ClosedRanges.UInt16Range, new SwiftClosedRange<ushort>(ushort.MinValue, ushort.MaxValue));
            Assert.Equal(ClosedRanges.UInt32Range, new SwiftClosedRange<uint>(uint.MinValue, uint.MaxValue));
            Assert.Equal(ClosedRanges.UInt64Range, new SwiftClosedRange<ulong>(ulong.MinValue, ulong.MaxValue));
            Assert.Equal(ClosedRanges.UIntRange, new SwiftClosedRange<nuint>(uint.MinValue, uint.MaxValue));
            Assert.Equal(ClosedRanges.Int8Range, new SwiftClosedRange<sbyte>(sbyte.MinValue, sbyte.MaxValue));
            Assert.Equal(ClosedRanges.Int16Range, new SwiftClosedRange<short>(short.MinValue, short.MaxValue));
            Assert.Equal(ClosedRanges.Int32Range, new SwiftClosedRange<int>(int.MinValue, int.MaxValue));
            Assert.Equal(ClosedRanges.Int64Range, new SwiftClosedRange<long>(long.MinValue, long.MaxValue));
            Assert.Equal(ClosedRanges.IntRange, new SwiftClosedRange<nint>(int.MinValue, int.MaxValue));
            Assert.Equal(ClosedRanges.FloatRange, new SwiftClosedRange<float>(-float.MaxValue, float.MaxValue));
            Assert.Equal(ClosedRanges.DoubleRange, new SwiftClosedRange<double>(-double.MaxValue, double.MaxValue));
            Assert.Equal(ClosedRanges.StringRange, new SwiftClosedRange<string>("A", "Z"));
        }

        [Fact]
        void testClosedRangeEcho() {
            Assert.Equal(ClosedRanges.EchoUInt8Range(ClosedRanges.UInt8Range), ClosedRanges.UInt8Range);
            Assert.Equal(ClosedRanges.EchoUInt16Range(ClosedRanges.UInt16Range), ClosedRanges.UInt16Range);
            Assert.Equal(ClosedRanges.EchoUInt32Range(ClosedRanges.UInt32Range), ClosedRanges.UInt32Range);
            Assert.Equal(ClosedRanges.EchoUInt64Range(ClosedRanges.UInt64Range), ClosedRanges.UInt64Range);
            Assert.Equal(ClosedRanges.EchoUIntRange(ClosedRanges.UIntRange), ClosedRanges.UIntRange);
            Assert.Equal(ClosedRanges.EchoInt8Range(ClosedRanges.Int8Range), ClosedRanges.Int8Range);
            Assert.Equal(ClosedRanges.EchoInt16Range(ClosedRanges.Int16Range), ClosedRanges.Int16Range);
            Assert.Equal(ClosedRanges.EchoInt32Range(ClosedRanges.Int32Range), ClosedRanges.Int32Range);
            Assert.Equal(ClosedRanges.EchoInt64Range(ClosedRanges.Int64Range), ClosedRanges.Int64Range);
            Assert.Equal(ClosedRanges.EchoIntRange(ClosedRanges.IntRange), ClosedRanges.IntRange);
            Assert.Equal(ClosedRanges.EchoFloatRange(ClosedRanges.FloatRange), ClosedRanges.FloatRange);
            Assert.Equal(ClosedRanges.EchoDoubleRange(ClosedRanges.DoubleRange), ClosedRanges.DoubleRange);
            Assert.Equal(ClosedRanges.EchoStringRange(ClosedRanges.StringRange), ClosedRanges.StringRange);

            // TODO: Empty ranges?
            // ClosedRange empty = new ClosedRange(0, -1)
            // Assert.True(empty.isEmpty())
            // Assert.True(ClosedRanges.EchoInt32Range(empty).isEmpty())
        }

        [Fact]
        void testBounds() {
            Assert.Equal(Ranges.UInt64Range.LowerBound, ulong.MinValue);
            Assert.Equal(Ranges.UInt64Range.UpperBound, ulong.MaxValue);
            Assert.Equal(ClosedRanges.UInt64Range.LowerBound, ulong.MinValue);
            Assert.Equal(ClosedRanges.UInt64Range.UpperBound, ulong.MaxValue);
        }
    }
}
